man/cryptsetup-tpm2.8
.TH CRYPTSETUP-TPM2 "8" "December 2019" "cryptsetup-tpm2" "Maintenance Commands"
.SH NAME
cryptsetup-tpm2 - tool for activating LUKS2 encrypted volumes using TPM2
.SH SYNOPSIS
.B cryptsetup-tpm2 <options> <action> <action args>
.SH DESCRIPTION
.PP
cryptsetup-tpm2 is used to conveniently lock and activate LUKS2 volumes using a
passphrase stored in a \fBTPM 2.0\fR. \fBTPM\fR is a secure cryptoprocessor
(similar to a smartcard) present in most of the recent laptops (and some desktops).
cryptsetup-tpm2 stores metadata in a \fItpm2\fR token inside the \fItokens\fR
section of the LUKS2 header (see \fILUKS2 On-Disk Format Specification\fR).

This utility supports sealing the TPM key to a current HW/SW state. To fully
utilize this feature, it is required that the state is measured to the PCR banks
by BIOS/UEFI, bootloader (e.g. \fITrustedGrub\fR) and kernel (IMA subsystem).
If anything in the boot chain deviates from the pre-measured state, the passphrase
won't be released from the TPM.

You can seal the passphrase to a specified PCR. They have the following meaning:
.IP
\fIPCR#0-7\fR Measured by BIOS/EFI. includes measurement of items like boot
options and order, microcode or secure boot status.
.IP
\fIPCR#16\fR Debug register

.PP
In TrustedGrub2 PCRs contain the following measurements:
.IP
\fIPCR#8\fR
First sector of TrustedGRUB2 kernel (diskboot.img)
.IP
\fIPCR#9\fR
TrustedGRUB2 kernel (core.img)
.IP
\fIPCR#10\fR
Loader measurements - currently linux-kernel, initrd, ntldr, chainloader, multiboot, module
.IP
\fIPCR#11\fR
Contains all commandline arguments from scripts (e.g. grub.cfg) and those entered in the shell
.IP
\fIPCR#12\fR
LUKS-header
.IP
\fIPCR#13\fR
Parts of GRUB2 that are loaded from disk like GRUB2-modules

.PP
To protect the TPM-stored passphrase \fIdictionary attack (DA) protection\fR
can be enabled inside the TPM. \fBBe careful not to enter wrong TPM password
too many times with this option enabled.\fR

\fBIt is strongly recommended to have a least one regular (non-TPM) passphrase
added.\fR This will prevent a lockout in case of changed PCR configuration
(e.g. after kernel update or kernel command line change) or after exceeded
number of attempts on a DA-protected passphrase.

.SH BASIC COMMANDS

\fIadd\fR <data_device> <options>
.IP
This will generate a random LUKS2 passphrase (using random data supplied from
the TPM) and store the passphrase inside NV memory of the TPM.

\fB<options>\fR can be [--tpm2-nv, --tpm2-pcr, --tpm2-bank, --tpm2-daprotect,
--tpm2-no-pin, --tpm2-key-size, --timeout=secs]


.PP
\fIopen\fR <data_device> <name>
.IP
Retrieves a LUKS2 passphrase from the TPM's NV index stored in the header
metadata. If the NV index is PIN protected, the user will be asked for the PIN.
Then \fB<data_device>\fR will be activated using the passphrase and opened as
\fB<name>\fR.

\fB<options>\fR can be [--token-id]

.PP
\fIkill\fR <data_device> --token-id <id> or --tpm2-nv=<nvindex>
.IP
This will remove the randomly generated passphrase from the TPM and the
associated LUKS2 keyslot. Specifying either token ID or NV index is mandatory.

.PP
\fIdump\fR <data_device>
.IP
The LUKS2 header will be dumped in a human-readable way. This includes the
fields specific to TPM2 token.

.PP
\fIlist\fR
.IP
This action does not operate on any data device. It only lists supported PCRs
for every bank.

.SH OPTIONS
.TP
.B --tpm2-nv=<0x01800000..0x01BFFFFF>
Specify an index on which to store the randomly generated passphrase if none is
supplied a free one is found.
.TP
.B --tpm2-pcr=<pcr>[,<pcr>[,<pcr>[...]]]
When supplied, the randomly generated passphrase will will be sealed to current
values of the specified PCRs.
.TP
.B --tpm2-bank=<hash1>[,<hash2>[,<hash3>[...]]]
Selection of banks for PCR sealing
.TP
.B --tpm2-daprotect
Enable dictionary attack protection on the TPM stored passphrase.
.TP
.B --tpm2-no-pin
Don't PIN-protect TPM NV index. This way you don't have to enter any
passphrase/PIN when activating the disk. \fBWhen enabled, it is strongly
recommended to have TPM enabled bootloader and kernel which make measurements
of the current state and to seal the passphrase to the appropriate PCRs. Even
then, the system is susceptible to cold-boot attacks.\fR
.TP
.B --tpm2-key-size=<bytes>
Size of a randomly generated passphrase which will be stored in the TPM. It
makes no sense to set it to a greater number than the master key length.
.TP
.B --token-id=INT
Specify the token number (as listed by the \fIlist\fR command)
.TP
.B --timeout=secs
Timeout for TPM PIN interactive passphrase prompt (in seconds)
.TP
.B --debug
Run in debug mode with full diagnostic logs. Debug output lines are always
prefixed by '#'.

.SH RETURN CODES
Cryptsetup returns 0 on success and a non-zero value on error.

Error codes are: 1 wrong parameters, 2 no permission (bad passphrase),
3 out of memory, 4 wrong device specified, 5 device already exists
or device is busy.

.SH AUTHORS
cryptsetup-tpm2 is based on code originally written by Andreas Fuchs from
Fraunhofer SIT sponsored by Infineon Technologies AG and further extended by
Daniel Zatovic <dan.zatovic@gmail.com>
.br
The cryptsetup-tpm2 man page was written by Daniel Zatovic
<dan.zatovic@gmail.com>
.br

.SH COPYRIGHT
Copyright \(co 2018-2019 Fraunhofer SIT sponsored by Infineon Technologies AG
.br
Copyright \(co 2019-2020 Red Hat, Inc. All rights reserved.
.br
Copyright \(co 2019-2020 Daniel Zatovic
.br
Copyright \(co 2009-2020 Milan Broz
.br

.B libcryptsetup

.IP
Copyright \(co 2004 Jana Saout
.br
Copyright \(co 2004-2006 Clemens Fruhwirth
.br
Copyright \(co 2012-2014 Arno Wagner
.br
Copyright \(co 2009-2020 Red Hat, Inc.
.br
Copyright \(co 2009-2020 Milan Broz
.br

.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH SEE ALSO
The LUKS website at \fBhttps://gitlab.com/cryptsetup/cryptsetup/\fR

The cryptsetup FAQ, contained in the distribution package and
online at
\fBhttps://gitlab.com/cryptsetup/cryptsetup/wikis/FrequentlyAskedQuestions\fR

The cryptsetup mailing list and list archive, see FAQ entry 1.6.

The LUKS on-disk format specification available at
\fBhttps://gitlab.com/cryptsetup/cryptsetup/wikis/Specification\fR


NEWS
See docs/* directory for Release Notes.


tokens/tpm2/cryptsetup-tpm2.c
/*
 * LUKS - Linux Unified Key Setup v2, TPM type keyslot handler
 *
 * Copyright (C) 2018-2020 Fraunhofer SIT sponsorred by Infineon Technologies AG
 * Copyright (C) 2019-2020 Red Hat, Inc. All rights reserved.
 * Copyright (C) 2019-2020 Daniel Zatovic
 * Copyright (C) 2019-2020 Milan Broz
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "../src/cryptsetup.h"
#include "utils_tpm2.h"

#define PACKAGE_CRYPTSETUP_TPM2 "cryptsetup-tpm2"

#define DEFAULT_PCR_BANK "sha256"
#define DEFAULT_TPM2_SIZE 64
#define DEFAULT_TPM2_SIZE_MAX 512

static uint32_t opt_tpmnv = 0;
static uint32_t opt_tpmpcr = 0;
static uint32_t opt_tpmbanks = 0;
static long int opt_pass_size = DEFAULT_TPM2_SIZE;
static int opt_tpmdaprotect = 0;
static int opt_no_tpm_pin = 0;
static int opt_token = CRYPT_ANY_TOKEN;
static int opt_timeout = 0;

static const char **action_argv;
static int action_argc;

void tools_cleanup(void) {}

static int action_tpm2_open(struct crypt_device *cd)
{
	int r;

	r = crypt_activate_by_token(cd, action_argv[1], opt_token, NULL, 0);

	return r < 0 ? r : 0;
}

static int action_tpm2_dump(struct crypt_device *cd)
{
	return crypt_dump(cd);
}

static int action_tpm2_kill(struct crypt_device *cd)
{
	const char *type;
	int i, r;

	if (!opt_tpmnv && opt_token == CRYPT_ANY_TOKEN) {
		l_err(cd, "Token ID or TPM2 nvindex option must be specified.");
		return -EINVAL;
	}

	if (opt_token == CRYPT_ANY_TOKEN)
		opt_token = tpm2_token_by_nvindex(cd, opt_tpmnv);

	if (opt_token < 0 ||
	    crypt_token_status(cd, opt_token, &type) != CRYPT_TOKEN_EXTERNAL ||
	    strcmp(type, "tpm2")) {
		l_err(cd, "No TPM2 token to destroy.");
		return -EINVAL;
	}

	/* Destroy all keyslots assigned to TPM 2 token */
	for (i = 0; i < crypt_keyslot_max(CRYPT_LUKS2) ; i++) {
		if (!crypt_token_is_assigned(cd, opt_token, i)) {
			r = crypt_keyslot_destroy(cd, i);
			if (r < 0) {
				l_err(cd, "Cannot destroy keyslot %d.", i);
				return r;
			}
		}
	}

	/* Destroy TPM2 NV index and token object itself */
	return tpm2_token_kill(cd, opt_token);
}

static int action_tpm2_add(struct crypt_device *cd)
{
	char *existing_pass = NULL, *tpm_pin = NULL, *random_pass = NULL;
	size_t existing_pass_len, tpm_pin_len = 0;
	int token, r, keyslot = CRYPT_ANY_SLOT;
	bool supports_algs_for_pcrs;
	TSS2_RC tpm_rc;

	if (!opt_tpmbanks) {
		l_err(cd, "PCR banks must be selected.");
		return -EINVAL;
	}

	tpm_rc = tpm2_supports_algs_for_pcrs(NULL, opt_tpmbanks, opt_tpmpcr, &supports_algs_for_pcrs);
	if (tpm_rc != TSS2_RC_SUCCESS) {
		l_err(NULL, "Failed to get PCRS capability from TPM.");
		LOG_TPM_ERR(NULL, tpm_rc);
		return -ECOMM;
	}

	if(!supports_algs_for_pcrs) {
		l_err(NULL, "Your TPM doesn't support selected PCR and banks combination.");
		return -ENOTSUP;
	}

	random_pass = crypt_safe_alloc(opt_pass_size);
	if (!random_pass)
		return -ENOMEM;

	r = tpm_get_random(cd, random_pass, opt_pass_size);
	if (r < 0)
		goto out;

	r = crypt_cli_get_key("Enter existing LUKS2 pasphrase:",
			  &existing_pass, &existing_pass_len,
			  0, 0, NULL, opt_timeout, 0, 0, cd, NULL);
	if (r < 0)
		goto out;

	if (!opt_no_tpm_pin) {
		r = crypt_cli_get_key("Enter new TPM password:",
				  &tpm_pin, &tpm_pin_len,
				  0, 0, NULL, opt_timeout, 1, 0, cd, NULL);
		if (r < 0)
			goto out;
	}

	if (opt_tpmnv == 0) {
		tpm_rc = tpm_nv_find(cd, &opt_tpmnv);
		if (tpm_rc != TSS2_RC_SUCCESS) {
			l_err(cd, "Error while trying to find free NV index.");
			LOG_TPM_ERR(cd, tpm_rc);
			r = -EINVAL;
			goto out;
		}

		if (!opt_tpmnv) {
			l_err(cd, "Error no free TPM NV-Index found.");
			r = -EACCES;
			goto out;
		}
	}

	tpm_rc = tpm_nv_define(cd, opt_tpmnv, tpm_pin, tpm_pin_len, opt_tpmpcr,
			  opt_tpmbanks, opt_tpmdaprotect, NULL, 0, opt_pass_size);
	if (tpm_rc != TSS2_RC_SUCCESS) {
		l_err(cd, "TPM NV-Index definition failed");
		LOG_TPM_ERR(cd, tpm_rc);
		r = -EINVAL;
		goto out;
	}

	tpm_rc = tpm_nv_write(cd, opt_tpmnv, tpm_pin, tpm_pin_len,
			random_pass, opt_pass_size);
	if (tpm_rc != TSS2_RC_SUCCESS) {
		l_err(cd, "TPM NV-Index write error.");
		LOG_TPM_ERR(cd, tpm_rc);
		tpm_nv_undefine(cd, opt_tpmnv);
		r = -EINVAL;
		goto out;
	}

	r = crypt_keyslot_add_by_passphrase(cd, keyslot, existing_pass, existing_pass_len, random_pass, opt_pass_size);
	if (r < 0) {
		if (r == -EPERM)
			l_err(cd, "Wrong LUKS2 passphrase supplied.");
		tpm_nv_undefine(cd, opt_tpmnv);
		goto out;
	}
	keyslot = r;
	l_std(cd, "Using keyslot %d.\n", keyslot);

	r = tpm2_token_add(cd, opt_tpmnv, opt_tpmpcr, opt_tpmbanks, opt_tpmdaprotect, !opt_no_tpm_pin, opt_pass_size);
	if (r < 0) {
		tpm_nv_undefine(cd, opt_tpmnv);
		crypt_keyslot_destroy(cd, keyslot);
		goto out;
	}
	token = r;
	l_std(cd, "Token: %d\n", token);

	r = crypt_token_assign_keyslot(cd, token, keyslot);
	if (r < 0) {
		l_err(cd, "Failed to assign keyslot %d to token %d.", keyslot, token);
		tpm_nv_undefine(cd, opt_tpmnv);
		crypt_keyslot_destroy(cd, keyslot);
		crypt_token_json_set(cd, token, NULL);
	}

	if (r > 0)
		r = 0;
out:
	crypt_safe_free(random_pass);
	crypt_safe_free(existing_pass);
	crypt_safe_free(tpm_pin);

	return r;
}

static int action_tpm2_list(struct crypt_device *cd)
{
	TSS2_RC r;
	TPMS_CAPABILITY_DATA *savedPCRs;
	TPMS_PCR_SELECTION *selection;
	unsigned i, j, k;

	r = getPCRsCapability(cd, &savedPCRs);
	if (r != TSS2_RC_SUCCESS) {
		l_err(cd, "Failed to get PCRS capability from TPM.");
		LOG_TPM_ERR(cd, r);
		return -EINVAL;
	}

	l_std(cd, "Supported PCRs for banks:\n");
	for (i = 0; i <  CRYPT_HASH_ALGS_COUNT; i++) {
		if ((selection = tpm2_get_pcrs_by_alg(savedPCRs, hash_algs[i].crypt_id))) {
			k = 0;
			l_std(cd, "%s:\t", hash_algs[i].name);
			for (j = 0; j < selection->sizeofSelect * 8; j++) {
				if (selection->pcrSelect[i/8] & (1 << i % 8)) {
					l_std(cd, "%s%d", k++ ? ", " : "", j);
				}
			}
			l_std(cd, "\n");
		}
	}

	free(savedPCRs);
	return 0;
}

static struct action_type {
	const char *type;
	int (*handler)(struct crypt_device*);
	int required_action_argc;
	const char *arg_desc;
	const char *desc;
} action_types[] = {
	{ "add",	action_tpm2_add,  1, N_("<data_device>"),N_("add TPM2 token") },
	{ "open",	action_tpm2_open, 2, N_("<data_device> <name>"),N_("open device as <name> using TPM token") },
	{ "kill",	action_tpm2_kill, 1, N_("<data_device> --token-id <id>"),N_("remove specified TPM token and related keyslots") },
	{ "dump",	action_tpm2_dump, 1, N_("<data_device>"),N_("show TPM2 token information") },
	{ "list",	action_tpm2_list, 0, N_("none"),N_("just list supported PCRs for banks") },
	{ NULL, NULL, 0, NULL, NULL }
};

static void help(poptContext popt_context,
		 enum poptCallbackReason reason __attribute__((unused)),
		 struct poptOption *key,
		 const char *arg __attribute__((unused)),
		 void *data __attribute__((unused)))
{
	struct action_type *action;
	unsigned i, n = 0;

	if (key->shortName == '?') {
		log_std("%s %s\n", PACKAGE_CRYPTSETUP_TPM2, PACKAGE_VERSION);
		poptPrintHelp(popt_context, stdout, 0);
		log_std(_("\n"
			 "<action> is one of:\n"));
		for(action = action_types; action->type; action++)
			log_std("\t%s %s - %s\n", action->type, _(action->arg_desc), _(action->desc));
		l_std(NULL, "\nDefault pass-size: %d [bytes]\n", DEFAULT_TPM2_SIZE);
		l_std(NULL, "Default PCRs: (none)\n");
		l_std(NULL, "Default PCR banks: %s\n", DEFAULT_PCR_BANK);

		l_std(NULL, "Possible PCR banks:");
		for (i = 0; i < CRYPT_HASH_ALGS_COUNT; i++) {
			printf("%s %s", n++ ? "," : "", hash_algs[i].name);
		}
		putchar('\n');

		exit(EXIT_SUCCESS);
	} else
		usage(popt_context, EXIT_SUCCESS, NULL, NULL);
}

static int run_action(struct action_type *action, struct crypt_device *cd)
{
	int r;

	log_dbg("Running command %s.", action->type);

	r = action->handler(cd);

	show_status(r);
	return translate_errno(r);
}

int main(int argc, const char **argv)
{
	int r = 0;
	struct crypt_device *cd = NULL;
	static const char *null_action_argv[] = {NULL};
	struct action_type *action;
	const char *aname;
	static long int tpmnv = 0;
	static const char *tpmbanks = DEFAULT_PCR_BANK;
	static const char *tpmpcr = NULL;

	static struct poptOption popt_help_options[] = {
		{ NULL,    '\0', POPT_ARG_CALLBACK, help, 0, NULL,                         NULL },
		{ "help",  '?',  POPT_ARG_NONE,     NULL, 0, N_("Show this help message"), NULL },
		{ "usage", '\0', POPT_ARG_NONE,     NULL, 0, N_("Display brief usage"),    NULL },
		POPT_TABLEEND
	};

	static struct poptOption token_popt_options[] = {
		{ NULL,             '\0', POPT_ARG_INCLUDE_TABLE, popt_help_options, 0, "Help options:", NULL },
		{ "tpm2-nv",        '\0', POPT_ARG_LONG,   &tpmnv,            0, "Select TPM's NV index", "<0x01800000..0x01BFFFFF>" },
		{ "tpm2-pcr",       '\0', POPT_ARG_STRING, &tpmpcr,       0, "Selection of TPM PCRs", "<pcr>[,<pcr>[,<pcr>[...]]]" },
		{ "tpm2-bank",      '\0', POPT_ARG_STRING, &tpmbanks,     0, "Selection of TPM PCR banks", "<hash1>[,<hash2>[,<hash3>[...]]]" },
		{ "tpm2-daprotect", '\0', POPT_ARG_NONE,   &opt_tpmdaprotect, 0, "Enable TPM dictionary attack protection", NULL},
		{ "tpm2-no-pin",    '\0', POPT_ARG_NONE,   &opt_no_tpm_pin,  0, "Don't PIN protect TPM NV index", NULL},
		{ "tpm2-key-size",  '\0', POPT_ARG_LONG,   &opt_pass_size,    0, "Size of randomly generated key for unlocking keyslot", "<bytes>"},
		{ "token-id",       '\0', POPT_ARG_INT,    &opt_token,        0, "Token number", NULL },
		{ "timeout",        '\0', POPT_ARG_INT,    &opt_timeout,      0, "Timeout for interactive passphrase prompt (in seconds)", "secs" },
		{ "debug",          '\0', POPT_ARG_NONE,   &opt_debug,        0, "Show debug messages", NULL },
		POPT_TABLEEND
	};
	poptContext popt_context;

	crypt_set_log_callback(NULL, tool_log, NULL);

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	popt_context = poptGetContext("cryptsetup_tpm2", argc, argv, token_popt_options, 0);
	poptSetOtherOptionHelp(popt_context, "[OPTION...] <action> <action-specific>");

	while ((r = poptGetNextOpt(popt_context)) > 0) ;
	if (r < -1)
		usage(popt_context, EXIT_FAILURE, poptStrerror(r),
		      poptBadOption(popt_context, POPT_BADOPTION_NOALIAS));

	if (!(aname = poptGetArg(popt_context)))
		usage(popt_context, EXIT_FAILURE, "Argument <action> missing.",
				poptGetInvocationName(popt_context));

	action_argc = 0;
	action_argv = poptGetArgs(popt_context);
	/* Make return values of poptGetArgs more consistent in case of remaining argc = 0 */
	if (!action_argv)
		action_argv = null_action_argv;

	while(action_argv[action_argc] != NULL)
		action_argc++;

	for (action = action_types; action->type; action++)
		if (strcmp(action->type, aname) == 0)
			break;

	if (!action->type)
		usage(popt_context, EXIT_FAILURE, _("Unknown action."),
		      poptGetInvocationName(popt_context));

	if (action_argc < action->required_action_argc) {
		char buf[128];
		snprintf(buf, 128,_("%s: requires %s as arguments"), action->type, action->arg_desc);
		usage(popt_context, EXIT_FAILURE, buf,
		      poptGetInvocationName(popt_context));
	}

	opt_tpmnv = (uint32_t)tpmnv;

	if (tpmpcr && tpm2_token_get_pcrs(tpmpcr, &opt_tpmpcr) < 0)
		usage(popt_context, EXIT_FAILURE, "Wrong PCR value.",
		      poptGetInvocationName(popt_context));

	if (tpmbanks && tpm2_token_get_pcrbanks(tpmbanks, &opt_tpmbanks) < 0)
		usage(popt_context, EXIT_FAILURE, "Wrong PCR bank value.",
				poptGetInvocationName(popt_context));

	if (opt_debug) {
		crypt_set_debug_level(CRYPT_DEBUG_ALL);
		dbg_version_and_cmd(argc, argv);
	}

	r = crypt_init(&cd, action_argv[0]);
	if (r < 0) {
		l_err(NULL, "Failed to init device %s.", argv[1]);
		return EXIT_FAILURE;
	}

	r = crypt_token_load(cd, "tpm2");
	if (r < 0) {
		l_err(cd, "Failed to load tpm2 token handler.");
		return EXIT_FAILURE;
	}

	if (strcmp(action->type, "list") != 0)
		r = crypt_load(cd, CRYPT_LUKS2, NULL);
	if (r < 0) {
		l_err(cd, "Failed to load luks2 device %s.", action_argv[0]);
		r = EXIT_FAILURE;
	} else
		r = run_action(action, cd);

	if (r) {
		l_err(cd, "Action %s FAILED.", action->type);
	} else {
		l_std(cd, "Action %s successful.\n", action->type);
	}

	crypt_free(cd);
	poptFreeContext(popt_context);
	return r;
}
